rfc/egress-traffic.md

```diff
@@ -0,0 +1,200 @@
+# RFC: Egress Traffic Tracking
+
+## Authors
+
+- Felipe Forbeck
+
+## Status
+
+- Proposed
+
+## Context
+
+Storacha Network aims to implement a scalable and automated mechanism to track egress traffic for customer billing purposes. The objective is to update Stripe’s API with relevant egress data for each customer, ensuring accurate billing. This RFC outlines two potential approaches to achieve this goal, each with its own set of trade-offs.
+
+## Problem
+
+Currently, there is no established automated process to track and associate egress traffic per customer and report this usage to Stripe for billing. A solution is required to automate this process for seamless, scalable, and accurate billing based on egress traffic.
+
+## Proposal
+
+We propose two alternatives for tracking and updating egress billing traffic:
+
+### Alternative 1: SQS, Lambda, DynamoDB, and Stripe Integration
+
+1. Introduce an SQS queue to receive egress event data from the Cloudflare Freeway Worker.
+2. Use a Lambda function to process these events and store the egress data in a DynamoDB table, mapping the data to the respective customer.
+3. The Lambda function will then update Stripe’s API with the corresponding egress data size for the customer.
+
+```mermaid
+graph TD
+    CF[Cloudflare Freeway Worker] --> SQS[SQS Queue]
+    SQS --> Lambda[Lambda Function]
+    Lambda --> DynamoDB[DynamoDB - Egress Data Storage]
+    Lambda --> Stripe[Stripe API]
+```
+
+**Advantages:**
+
+- Decoupled, event-driven architecture using SQS and Lambda, providing clear separation of responsibilities.
+- Fine-grained control over event handling and processing.
+
+**Disadvantages:**
+
+- Requires managing additional infrastructure (SQS queue, Lambda function, DynamoDB table) solely for this purpose.
+- Cloudflare Freeway Worker must be authorized to publish events to the SQS queue, adding complexity to authentication and authorization mechanisms.
+- Increased operational overhead due to the need to monitor and maintain new infrastructure components.
```

Comment:
Stripe does seem to be advertising the new "billing meters" that we'll need to migrate to sooner or later as something that can "scale right along with us" - we are still fairly small fish compared to their biggest users so it seems reasonable to hope this will support the kind of volume we want to send at it! 

I'd propose we do the first implementation with the queue-based approach in option 3 and worry about rolling up usage charges ourself if Stripe starts really buckling under the load:

https://stripe.com/resources/more/what-is-metered-billing-heres-how-this-adaptable-billing-model-works
https://stripe.com/blog/three-ways-our-usage-based-billing-product-is-unique

Reply:
From their docs:
> Increased rate limits up to 100,000 events per second

Anyways, the first iteration is using a queue and we can fine-tune that as needed.